File: javascript/rpc/caller.ts
import * as internal from "./internal.ts";

export interface Caller {
    call(method: string, args: any): Promise<internal.Response>;
}

export class Call {
    Destination: string;
    objectPath: string;
    method: string;
    caller: Caller|undefined;
    decode: any;

    constructor(destination: string) {
        this.Destination = destination;
        if (this.Destination.length === 0) {
            throw new Error("no destination specified");
        }
        if (this.Destination[0] == "/") {
            this.Destination = this.Destination.substr(1);
        }
        var parts = this.Destination.split("/");
        if (parts.length === 1) {
            this.objectPath = "/";
            this.method = parts[0];
            return;
        }
        this.method = parts.pop() as string;
        this.objectPath = parts.join("/");
    }

}

export class Response {
    error: string|undefined;
    hijacked: boolean;
    reply: any;
    channel: internal.IChannel;

    constructor(channel: internal.IChannel) {
        this.channel = channel;
        this.error = undefined;
        this.hijacked = false;
    }
}

export class caller implements Caller {
    session: internal.ISession;
    codec: internal.Codec;

    constructor(session: internal.ISession, codec: internal.Codec) {
        this.session = session;
        this.codec = codec;        
    }

    async call(path: string, args: any): Promise<internal.Response> {
        try {
            let ch = await this.session.open();
            let resp: internal.Response = new internal.Response(ch);
            let codec = new internal.FrameCodec(ch, this.codec, 2);
            
            await codec.encode(new Call(path));
            await codec.encode(args);

            let header: internal.ResponseHeader = await codec.decode();
            if (header.Error !== undefined && header.Error !== null) {
                // await ch.close();
                // return Promise.reject(header.Error);
                console.error(header);
            }
            
            resp.error = header.Error;
            resp.hijacked = header.Hijacked;
            resp.reply = await codec.decode();
            if (resp.hijacked !== true) {
                await ch.close();
            }
            
            return resp;
        } catch (e) {
            console.error(e, path, args);
            return Promise.reject("call?");
            //await ch.close();
        }
    }
}

File: golang/rpc/caller.go
package rpc

import (
	"context"
	"errors"
	"fmt"
	"io"
	"strings"

	"github.com/manifold/qtalk/golang/mux"
)

type Caller interface {
	Call(path string, args, reply interface{}) (*Response, error)
}

type Response struct {
	ResponseHeader

	Reply   interface{}
	Channel mux.Channel
}

type Call struct {
	Destination string
	ObjectPath  string
	Method      string
	Caller      Caller
	Decoder     Decoder
	Context     context.Context
}

func (c *Call) Parse() error {
	if len(c.Destination) == 0 {
		return fmt.Errorf("no destination specified")
	}
	if c.Destination[0] == '/' {
		c.Destination = c.Destination[1:]
	}
	parts := strings.Split(c.Destination, "/")
	if len(parts) == 1 {
		c.ObjectPath = "/"
		c.Method = parts[0]
		return nil
	}
	c.ObjectPath = strings.Join(parts[0:len(parts)-1], "/")
	c.Method = parts[len(parts)-1]
	return nil
}

func (c *Call) Decode(v interface{}) error {
	return c.Decoder.Decode(v)
}

type caller struct {
	session mux.Session
	codec   Codec
}

func NewCaller(session mux.Session, codec Codec) Caller {
	return &caller{
		session: session,
		codec:   codec,
	}
}

func (c *caller) Call(path string, args, reply interface{}) (*Response, error) {
	ch, err := c.session.Open()
	if err != nil {
		return nil, err
	}

	codec := &frameCodec{c.codec}
	enc := codec.Encoder(ch)
	dec := codec.Decoder(ch)

	// request
	err = enc.Encode(Call{
		Destination: path,
	})
	if err != nil {
		ch.Close()
		return nil, err
	}

	err = enc.Encode(args)
	if err != nil {
		ch.Close()
		return nil, err
	}

	// response
	var header ResponseHeader
	err = dec.Decode(&header)
	if err != nil {
		ch.Close()
		return nil, err
	}

	if !header.Hijacked {
		defer ch.Close()
	}

	resp := &Response{
		ResponseHeader: header,
		Channel:        ch,
		Reply:          reply,
	}
	if resp.Error != nil {
		return resp, fmt.Errorf("remote: %s", *(resp.Error))
	}

	if reply == nil {
		// read into throwaway buffer
		var buf []byte
		if err := dec.Decode(&buf); err != nil {
			if errors.Is(err, io.EOF) {
				return resp, nil
			}
			return resp, err
		}
	} else {
		if err := dec.Decode(resp.Reply); err != nil {
			return resp, err
		}
	}

	return resp, nil
}
